internal/unionstore/art/node.go

```diff
@@ -0,0 +1,741 @@
+package art
+
+import (
+	"bytes"
+	"math"
+	"math/bits"
+	"unsafe"
+
+	"github.com/tikv/client-go/v2/kv"
+)
+
+type nodeKind uint16
+
+const (
+	typeInvalid nodeKind = 0
+	typeNode4   nodeKind = 1
+	typeNode16  nodeKind = 2
+	typeNode48  nodeKind = 3
+	typeNode256 nodeKind = 4
+	typeLeaf    nodeKind = 5
+)
+
+const (
+	maxPrefixLen = 10
+	node4cap     = 4
+	node16cap    = 16
+	node48cap    = 48
+	node256cap   = 256
+)
+
+const (
+	// if we use 16-bit idx for nodeAddr, and break nodeAddr into embed fields, the artNodeSize can be 8 bits.
+	// By reducing 4 bits for nodeAddr, we can save about 1/4-1/3 memories for node.
+	artNodeSize = 12
+	node4size   = 80
+	node16size  = 236
+	node48size  = 888
+	node256size = 3096
+	leafSize    = 12
+)
+
+var (
+	// nullNode48 is used to initialize the node 256
+	nullNode48 = node48{}
+	// nullNode256 is used to initialize the node 256
+	nullNode256 = node256{}
+)
+
+var nullArtNode = artNode{kind: typeInvalid, addr: nullAddr}
+
+func init() {
+	for i := 0; i < node48cap; i++ {
+		nullNode48.children[i] = artNode{kind: typeInvalid, addr: nullAddr}
+	}
+	for i := 0; i < node256cap; i++ {
+		nullNode256.children[i] = artNode{kind: typeInvalid, addr: nullAddr}
+	}
+}
+
+type Key []byte
+type artNode struct {
+	kind nodeKind
+	addr nodeAddr
+}
+
+type node struct {
+	nodeNum     uint8
+	prefixLen   uint8
+	prefix      [maxPrefixLen]byte
+	inplaceLeaf artNode
+}
+
+type node4 struct {
+	node
+	present  uint8 // bitmap, only lower 4-bit is used
+	keys     [node4cap]byte
+	children [node4cap]artNode
+}
+
+type node16 struct {
+	present uint16 // bitmap
+	node
+	keys     [node16cap]byte
+	children [node16cap]artNode
+}
+
+// Node with 48 children
+const (
+	n48s = 6  // 2^n48s == n48m
+	n48m = 64 // it should be sizeof(node48.present[0])
+)
+
+type node48 struct {
+	node
+	present  [4]uint64
+	keys     [node256cap]uint8 // map byte to index
+	children [node48cap]artNode
+}
+
+type node256 struct {
+	node
+	children [node256cap]artNode
+}
+
+type leaf struct {
+	vAddr nodeAddr
+	klen  uint16
+	flags uint16
+}
+
+func (n *artNode) isLeaf() bool {
+	return n.kind == typeLeaf
+}
+
+func (n *artNode) leaf(a *artAllocator) *leaf {
+	return a.getLeaf(n.addr)
+}
+
+func (n *artNode) node(a *artAllocator) *node {
+	switch n.kind {
+	case typeNode4:
+		return &n.node4(a).node
+	case typeNode16:
+		return &n.node16(a).node
+	case typeNode48:
+		return &n.node48(a).node
+	case typeNode256:
+		return &n.node256(a).node
+	default:
+		panic("invalid node kind")
+	}
+}
+
+func (n *artNode) at(a *artAllocator, idx int) artNode {
+	switch n.kind {
+	case typeNode4:
+		return n.node4(a).children[idx]
+	case typeNode16:
+		return n.node16(a).children[idx]
+	case typeNode48:
+		n48 := n.node48(a)
+		return n48.children[n48.keys[idx]]
+	case typeNode256:
+		return n.node256(a).children[idx]
+	default:
+		panic("invalid node kind")
+	}
+}
+
+func (n4 *node4) init() {
```

Comment:
Can we somehow reduce duplicated code in all `init`s?
Or can we consider using generics to reduce more duplicated code? Maybe a benchmark is needed.